velox/expression/ComplexProxyTypes.h

```diff
@@ -0,0 +1,220 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <iterator>
+#include <optional>
+#include "velox/common/base/Exceptions.h"
+#include "velox/core/CoreTypeSystem.h"
+#include "velox/vector/TypeAliases.h"
+#include "velox/vector/VectorTypeUtils.h"
+
+namespace facebook::velox::exec {
+
+template <typename T>
+struct VectorReader;
+
+template <typename T, typename B>
+struct VectorWriter;
+
+// Lightweight object that can be used as a proxy for array primitive elements.
+// It is returned by ArrayProxy::operator()[].
+template <typename T>
+struct PrimitiveWriterProxy {
+  using vector_t = typename TypeToFlatVector<T>::type;
+  using element_t = typename CppToType<T>::NativeType;
+
+  PrimitiveWriterProxy(vector_t* flatVector, vector_size_t index)
+      : flatVector_(flatVector), index_(index) {}
+  void operator=(std::nullopt_t) {
+    flatVector_->setNull(index_, true);
+  }
+
+  void operator=(element_t value) {
+    flatVector_->set(index_, value);
+  }
+
+  void operator=(const std::optional<element_t>& value) {
+    if (value.has_value()) {
+      flatVector_->set(index_, value);
+    } else {
+      flatVector_->setNull(index_, true);
+    }
+  }
+
+ private:
+  vector_t* flatVector_;
+  vector_size_t index_;
+};
+
+// The object passed to the simple function interface that represent a single
+// array entry.
+// ## General Interface:
+// - add_item()  : Add not null item and return proxy to the value to be
+// written.
+// - add_null()  : Add null item.
+// - size()      : Return the size of the array.
+
+// ## Special std::like interfaces when V is primitive:
+// - resize(n)         : Resize to n, nullity not written.
+// - operator[](index) : Returns PrimitiveWriterProxy which can be used to write
+// value or nullity at index.
+// - push_back(std::optional<v> value) : Increase size by 1, adding a value or
+// null.
+template <typename V>
+class ArrayProxy {
+  using child_writer_t = VectorWriter<V, void>;
+  using element_t = typename child_writer_t::exec_out_t;
+
+ public:
+  ArrayProxy<V>(const ArrayProxy<V>&) = delete;
+
+  // String and bool not yet supported, this probably wont work for string.
+  static bool constexpr provide_std_interface = CppToType<V>::isPrimitiveType &&
+      !std::is_same<Varchar, V>::value && !std::is_same<bool, V>::value;
+
+  // Add a new not null item to the array, increasing its size by 1.
+  FOLLY_ALWAYS_INLINE element_t& add_item() {
```

Comment:
nit: wondering if append or push_back might be more intuitive

Reply:
I dont have strong opinion here between add_item() and append(), I can use whatever you prefer @mbasmanova 
I would avoid push_back() not to confuse with the std::like one void push_back(v).

Reply:
if we use append
I would change add_ item to append_null()

Reply:
+1 let's avoid push_back() and emplace_back() since those methods normally return void.

Could we have users use the std-lib compatible pattern
```
a.emplace_back(); // Could forward the arguments if that is useful (e.g. for initial array size).
auto& element = a.back();
```
however while this is compatible, it's not natural. `auto& element = a.add_item();` is a little bit nicer IMO.